test/odoc_print/odoc_print.ml

```diff
@@ -8,26 +8,192 @@ let print_json_desc desc x =
   let yojson = Type_desc_to_yojson.to_yojson desc x in
   Yojson.Basic.pretty_print Format.std_formatter yojson
 
-let run inp =
+module Element = struct
+  open Odoc_model.Lang
+
+  type t =
+    | Module of Module.t
+    | ModuleType of ModuleType.t
+    | Type of TypeDecl.t
+    | Value of Value.t
+end
+
+let rec signature_of_simple_expansion :
+    Odoc_model.Lang.ModuleType.simple_expansion -> Odoc_model.Lang.Signature.t =
+  function
+  | Signature sg -> sg
+  | Functor (_, e) -> signature_of_simple_expansion e
+
+and signature_of_module_type_expr = function
+  | Odoc_model.Lang.ModuleType.Signature sg -> Some sg
+  | Path { p_expansion = Some exp; _ } ->
+      Some (signature_of_simple_expansion exp)
+  | Path { p_expansion = None; _ } -> None
+  | Functor (_, e) -> signature_of_module_type_expr e
+  | TypeOf { t_expansion = Some e; _ } -> Some (signature_of_simple_expansion e)
+  | TypeOf _ -> None
+  | With { w_expansion = Some e; _ } -> Some (signature_of_simple_expansion e)
+  | With _ -> None
+
+and signature_of_module :
+    Odoc_model.Lang.Module.t -> Odoc_model.Lang.Signature.t option =
+ fun m ->
+  match m.type_ with
+  | Alias (_, Some e) -> Some (signature_of_simple_expansion e)
+  | Alias (_, None) -> None
+  | ModuleType m -> signature_of_module_type_expr m
+
+and signature_of_module_type :
+    Odoc_model.Lang.ModuleType.t -> Odoc_model.Lang.Signature.t option =
+ fun m ->
+  match m.expr with Some e -> signature_of_module_type_expr e | None -> None
+
+let find_map fn list =
+  let rec inner = function
+    | x :: xs -> ( match fn x with Some y -> Some y | None -> inner xs)
+    | [] -> None
+  in
+  inner list
+
+let find_module name sg =
+  let open Odoc_model.Lang.Signature in
+  find_map
+    (function
+      | Module (_, ({ id; _ } as m))
+        when Odoc_model.Paths.Identifier.name id = name ->
+          Some (Element.Module m)
+      | _ -> None)
+    sg.items
+
+let find_module_type name sg =
+  let open Odoc_model.Lang.Signature in
+  find_map
+    (function
+      | ModuleType ({ id; _ } as m)
+        when Odoc_model.Paths.Identifier.name id = name ->
+          Some (Element.ModuleType m)
+      | _ -> None)
+    sg.items
+
+let find_type name sg =
+  let open Odoc_model.Lang.Signature in
+  find_map
+    (function
+      | Type (_, ({ id; _ } as m))
+        when Odoc_model.Paths.Identifier.name id = name ->
+          Some (Element.Type m)
+      | _ -> None)
+    sg.items
+
+let find_value name sg =
+  let open Odoc_model.Lang.Signature in
+  find_map
+    (function
+      | Value ({ id; _ } as m) when Odoc_model.Paths.Identifier.name id = name
+        ->
+          Some (Element.Value m)
+      | _ -> None)
+    sg.items
+
+(* Really cut-down reference lookup! *)
+let rec handle_ref :
```

Comment:
A simpler implementation might be to open the signature into an empty environment and use the env instead of implementing lookups.